CLDSRV-333: Handle delete and overwrite of MPUs

dvasilas opened the pull request:
When a MPU is deleted or overwritten, send its `uploadId` to the metadata backend.
This information is used to cleanup the replay key associated with the MPU.  

This PR is related to https://github.com/scality/MetaData/pull/2097

Issue: CLDSRV-333

dvasilas committed:
CLDSRV-333: objectDelete: add MPU uploadId to delOptions

If the object to be deleted is a MPU,
include its uploadId as a field named replayId
to the options with which services.deleteObject
is called.
Metadata will use this field to cleanup the replay
key corresponding to the overwriten MPU.

dvasilas committed:
CLDSRV-333: add oldReplayId to metadataStoreParams on overwrite

When a MPU is overwritten by a PUT (createAndStoreObject)
or a copy (objectCopy), add the MPU's uploadId to
metadataStoreParams in a field called oldReplayId.
Metadata will use this field to cleanup the replay
key corrensponding to the overwritten MPU.

dvasilas committed:
CLDSRV-333: add mpuUtils

Avoid code duplication in tests that create MPUs.

dvasilas committed:
CLDSRV-333: test MPU delete/overwrite handling.

Add unit tests that check that the correct
fields are sent to Metadata when a MPU is
deleted or overwritten.

dvasilas committed:
CLDSRV-333: catch unintentional error

bert-e commented:
# Hello dvasilas,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-333 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.10.25`

* `7.70.16`

* `8.4.20`

* `8.5.21`

* `8.6.8`

* `8.7.6`


Please check the `Fix Version/s` of CLDSRV-333, or the target
branch of this pull request.

dvasilas commented:
:robot:

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-333 contains:


* `7.10.16`

* `7.10.25`

* `8.4.20`

* `8.5.21`

* `8.6.8`

* `8.7.6`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.10.25`

* `7.70.16`

* `8.4.20`

* `8.5.21`

* `8.6.8`

* `8.7.6`


Please check the `Fix Version/s` of CLDSRV-333, or the target
branch of this pull request.

dvasilas commented:
:robot:

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.4/improvement/CLDSRV-333-handle-MPU-delete-overwrite` with contents from `w/7.70/improvement/CLDSRV-333-handle-MPU-delete-overwrite`
and `development/8.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.4/improvement/CLDSRV-333-handle-MPU-delete-overwrite origin/development/8.4
 $ git merge origin/w/7.70/improvement/CLDSRV-333-handle-MPU-delete-overwrite
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/improvement/CLDSRV-333-handle-MPU-delete-overwrite
```

dvasilas commented:
:robot:

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.5/improvement/CLDSRV-333-handle-MPU-delete-overwrite` with contents from `w/8.4/improvement/CLDSRV-333-handle-MPU-delete-overwrite`
and `development/8.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.5/improvement/CLDSRV-333-handle-MPU-delete-overwrite origin/development/8.5
 $ git merge origin/w/8.4/improvement/CLDSRV-333-handle-MPU-delete-overwrite
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.5/improvement/CLDSRV-333-handle-MPU-delete-overwrite
```

dvasilas commented:
:robot:

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-333-handle-MPU-delete-overwrite` into
`development/7.10`
* [w/7.70/improvement/CLDSRV-333-handle-MPU-delete-overwrite](https://github.com/scality/cloudserver/compare/development/7.70...w/7.70/improvement/CLDSRV-333-handle-MPU-delete-overwrite) will be merged into `development/7.70` 
* [w/8.4/improvement/CLDSRV-333-handle-MPU-delete-overwrite](https://github.com/scality/cloudserver/compare/development/8.4...w/8.4/improvement/CLDSRV-333-handle-MPU-delete-overwrite) will be merged into `development/8.4` 
* [w/8.5/improvement/CLDSRV-333-handle-MPU-delete-overwrite](https://github.com/scality/cloudserver/compare/development/8.5...w/8.5/improvement/CLDSRV-333-handle-MPU-delete-overwrite) will be merged into `development/8.5` 
* [w/8.6/improvement/CLDSRV-333-handle-MPU-delete-overwrite](https://github.com/scality/cloudserver/compare/development/8.6...w/8.6/improvement/CLDSRV-333-handle-MPU-delete-overwrite) will be merged into `development/8.6` 
* [w/8.7/improvement/CLDSRV-333-handle-MPU-delete-overwrite](https://github.com/scality/cloudserver/compare/development/8.7...w/8.7/improvement/CLDSRV-333-handle-MPU-delete-overwrite) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

WillToozsScality reviewed (approved):
lgtm

dvasilas committed:
CLDSRV-333 - Make docker image compatible with scality's registry

Images built with the `provenance` flag set to `true`
are currently not supported by our registry and older versions of docker.

dvasilas commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/cloudserver/commit/95c180e9d908375dccd0c5c1dd1c1ddeb816645c) did not succeed in branch w/8.4/improvement/CLDSRV-333-handle-MPU-delete-overwrite.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/cloudserver/commit/624d4708cf86f6b9e9b1b03abde99242bf53ab5e) did not succeed in branch w/8.5/improvement/CLDSRV-333-handle-MPU-delete-overwrite.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.4`



Please check the status of the associated issue CLDSRV-333.

Goodbye dvasilas.
